apps/service_providers/llm_service/adapters.py

```diff
@@ -0,0 +1,339 @@
+        # TODO: change this to something specific to the current chat message
+        return self.session.chat.metadata.get("cancelled", False)
+
+
+class AssistantAdapter(BaseAdapter):
+    def __init__(
+        self,
+        session: ExperimentSession,
+        assistant: OpenAiAssistant,
+        citations_enabled: bool,
+        experiment_version_number: int,
+        experiment_is_a_version: bool,
+        input_formatter: str | None = None,
+        trace_service=None,
+        save_message_metadata_only: bool = False,
+    ):
+        self.session = session
+        self.assistant = assistant
+        self.llm_service = assistant.llm_provider.get_llm_service()
+        self.citations_enabled = citations_enabled
+        self.experiment_version_number = experiment_version_number
+        self.experiment_is_a_version = experiment_is_a_version
+        self.input_formatter = input_formatter
+        self.trace_service = trace_service
+        self.save_message_metadata_only = save_message_metadata_only
+
+        self.provider_model_name = assistant.llm_provider_model.name
+        self.input_message_metadata = {}
+        self.output_message_metadata = {}
+        self.team = session.team
+
+        self.tools = get_assistant_tools(assistant, experiment_session=session)
+        self.template_context = PromptTemplateContext(session, source_material_id=None)
+
+    @staticmethod
+    def for_experiment(experiment: Experiment, session: ExperimentSession, trace_service=None) -> Self:
+        return AssistantAdapter(
+            session=session,
+            assistant=experiment.assistant,
+            citations_enabled=experiment.citations_enabled,
+            experiment_version_number=experiment.version_number,
+            experiment_is_a_version=experiment.is_a_version,
+            input_formatter=experiment.input_formatter,
+            trace_service=trace_service,
+        )
+
+    @staticmethod
+    def for_pipeline(session: ExperimentSession, node: "AssistantNode") -> Self:
+        assistant = OpenAiAssistant.objects.get(id=node.assistant_id)
+        experiment = session.experiment
+        return AssistantAdapter(
+            session=session,
+            assistant=assistant,
+            citations_enabled=node.citations_enabled,
+            experiment_version_number=experiment.version_number,
+            experiment_is_a_version=experiment.is_a_version,
+            input_formatter=node.input_formatter,
+            trace_service=experiment.trace_service,
+            save_message_metadata_only=True,
+        )
+
+    @cached_property
+    def assistant_client(self):
+        return self.assistant.get_assistant().client
+
+    @property
+    def assistant_tools_enabled(self):
+        return self.assistant.tools_enabled
+
+    @property
+    def assistant_builtin_tools(self) -> list:
+        return self.assistant.builtin_tools
+
+    @property
+    def thread_id(self):
+        return self.session.chat.thread_id
```

Comment:
I think this is causing the test failures:

`AttributeError: 'Chat' object has no attribute 'thread_id'`